### ISSUE TITLE
Capture scaled monitors

### ISSUE THREAD

**Issue by vicnevicne** [Ginj-capture/Ginj#15]
@janborup wrote:

I have 3 monitors 
Monitor 1: 1920x1080 Scaling : 125%
Monitor 2: 1680x1050 Scaling : 100%
Monitor 3: 1680x1050 Scaling : 100%

Normal all monitor, look likes this:
![image](https://user-images.githubusercontent.com/6866443/87929113-33b2b680-ca86-11ea-9d95-cbbb47c2b4c6.png)

When i start Ginj capture, Ginj up-scaling all screens cracy, the bottom incl. taskbar dissapears, and the text look pixelated:
![image](https://user-images.githubusercontent.com/6866443/87929216-5ba21a00-ca86-11ea-92fe-f5a98046816d.png)

When I mark/drag a window, example the calculator on monitor 2, Ginj select something different:
![image](https://user-images.githubusercontent.com/6866443/87929581-f00c7c80-ca86-11ea-9c8d-00d6b5fca9b9.png)

The above screenshot is with Windows standard capture, that is working fine with the multible monitors.

Is this fixable ?

/Cheers

**Comment by vicnevicne** [comment]
I thought it was unrelated to the fact of having multiple monitors, but now I'm not sure.
I only have a laptop with me (and no external screen), and performed tests at 100% and 125% with no issue, even starting GInj, then changing scaling, then capturing seems to work fine.
So now I'm wondering if, indeed, the issue is not due to combining scaling and multiscreen in the end...
Could you please check if everything works OK if you configure Windows to only display on the primary screen ("PC screen only") and it is scaled at 125% ?
And also, could you check if everything works OK if you configure Windows on the 3 screens, but all scaled at 100%
Finally, can you confirm the primary is the one on the left ?

**Comment by janborup** [comment]
@vicnevicne 
I'm on a little vacation, so when I'm back I will give you a detailed report on every scenario of this issue.
Keep chilling 😎

**Comment by vicnevicne** [comment]
Hey ! You were fast :-)
Indeed, I haven't (attempted to) fix scaling issues in 0.4.1, and I guess nothing will work until I figure out the coordinate change required to correctly handle the second screen.

But I just performed a few tests, with an HD TV connected as secondary display, and the results reported by Java are:

- With TV to the right of main display :
@100% (reference)
Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1920x1080@2560,0 minus 0/0/0/0 insets.
@125% 
Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1536x864@2048,0 minus 0/0/0/0 insets.
@150% 
Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1280x720@1707,0 minus 0/0/0/0 insets

- With TV "above" :
@150% Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1280x720@0,-720 minus 0/0/0/0 insets.

- With TV "below" :
@150% Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1280x720@0,960 minus 0/0/0/0 insets

What is very surprising is that 2048: 
Screen.config 0.0: 2560x1440@0,0 minus 0/0/0/0 insets.
Screen.config 1.0: 1536x864@**_2048,0_** minus 0/0/0/0 insets.
(tv to the right, 125%)

In other words, the same coordinates represent 2 different places. 
E.g. The point at (2348, 0) can mean either x=2348 on main screen or x = 300 (2348 - 2048) on the secondary screen
Now of course when Java returns me the mouse position in terms of an (x,y) point, I have no way to know which actual point it represents.
Hmm, there must be a way to handle that... 

Btw, as I understood you have C# experience, would you have time to quickly wrap up a C# app that dumps mouse coordinates on the console and see if those weird coordinates are native from Windows or if it's the result of a bad conversion by Java ?

**Comment by janborup** [comment]
Hi Again @vicnevicne 

> ".. One-Monitor 125% scaled.. "
Just take a look at my last post, here is the same problem with only One scaled monitor with the Video-capture.... so here os no second+ screen in the game.

I will find time to return the mouse-pos, and give you a report.

**Comment by vicnevicne** [comment]
> One-Monitor
Oh yes, missed that bit, sorry.
OK, so this might be another issue because I'm grabbing the coordinates from Java and then passing them to ffmpeg untouched, and maybe Java interprets the scaling differently from ffmpeg. I'll have to do more tests, but I admit video is coming second in my priority list. Scaling apart, I haven't tested video capture with multi-display either.
Thanks for the tests and reports.

**Comment by janborup** [comment]
I think the problem could be, in the second you click the "capture" (before dragging a window), because then seem like the whole screen is doing some screen scaling/antialiasing/bluring - so maybe if you got coordinates in this mode, it's intetterpered wrong some how.

**Comment by vicnevicne** [comment]
Yes. the way Ginj (and Jing) work is that indeed at the exact time you click the "Capture" button, a "megascreenshot" is performed, so that the state is "frozen". Then a "megawindow" is opened and displayed in "full-multi-screen", with the result of the capture as background. Finally, the yellow selection crossbar is displayed over it.
In other words, the area selection is not performed on the screen, it is performed on an pre-made screenshot, and finalizing the capture is just a clipping.
And you're right, that's when the coordinates issue happens for sure.

**Comment by janborup** [comment]
Hi @vicnevicne 

Now i got the time to create at MousePos-app an here is the result at at the corners of the screen..

![image](https://user-images.githubusercontent.com/6866443/92455925-912fcd80-f1c2-11ea-8d7a-e03388add1a2.png)

Observe.. 
on the first 125% screen, this resolution is 1530x864, but second screen still starts at X:1920
on the first 100% screen, this resolution is 1920x1080, and second screen still starts at X:1920

Hope this clarify the problem.

**Comment by vicnevicne** [comment]
Hi Jan,
Thanks for the test. I guess that is with a C++ or C# app, right ?
These last few days, I tried to to a similar thing with a dedicated Java app, and posted my findings (and question) [here](https://stackoverflow.com/questions/63770033/java-multi-display-handling-under-windows-bug-with-scaled-displays).
To summarize, using the Java API, I am getting correctly scaled Windows, but it looks like the coordinates of the top left corner are also "scaled" which makes no sense of course, and causes "logical" screen areas to overlap.
Now I have tested the JNA approach (JNA is a way to call native APIs from Java) but it seems it returns physical information (scaling has no effect on the returned values).
Now it seems all this depend how the (native) app declares itself to Windows on startup (using the manifest) - https://docs.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows - No idea how Java declares itself...
Could you please share the code you used ? I might adapt it to JNA and see if I'm getting the same values as you

**Comment by janborup** [comment]
Code !!!

No fancy... only 2 linjes of code.

[StackOverflow](https://stackoverflow.com/questions/8201286/get-cursor-position-with-respect-to-the-control-c-sharp/63795263#63795263)

**Comment by vicnevicne** [comment]
Indeed :-). But it's C# and JNA links to C++ libs.
My current JNA code calls User32's GetCursorPos(point) function, but I'm not sure how different it is from your C# code. In other words I don't know if the difference lies between GetCursorPos and MousePosition, or between an JNA and a native app.
Anyway, I think I have all the info I need by combining Java and JNA. Now I just need to figure out how to make good use of it.
Unfortunately, day job has taken over, so I need a bit of time ;-). I'll post progress here of course. Thanks for your help.

**Comment by janborup** [comment]
Just created an app to tell the Width * Height of all screens... 

![image](https://user-images.githubusercontent.com/6866443/92617380-8444d400-f2bf-11ea-9c0b-362ea91cfb71.png)

[MousePos.zip](https://github.com/Ginj-capture/Ginj/files/5196115/MousePos.zip)

**Comment by vicnevicne** [comment]
OK, Confirmed by a StackOverflow user as being a Java bug - https://bugs.openjdk.java.net/browse/JDK-8211999 - I'll try to mix in some JNA...

**Comment by vicnevicne** [comment]
Hi Jan, hi all,
I wrapped up v0.4.2pre with a JNA hack that hopefully fixes the issue with multiple displays and scaling. 
There is still a known visual artifact when selecting an area in the "overlapping zone" that sometimes creates a non-dimmed rectangle to appear outside of the selection, but I think it's minor (and not an easy fix), so if capture works ok and you don't find any regression (don't hesistate to shake it up :-) ), I intend to release it as 0.4.2.
In the meantime, as it is tagged "prerelease", it will not update automatically so you'll have to download the exe from https://github.com/Ginj-capture/Ginj/releases/tag/v0.4.2-pre to install it (no need to uninstall first).
All feedback is welcome.

**Comment by vicnevicne** [comment]
Hi all,
I just released v0.4.2 with an improved hack (compared to v0.4.2pre) which should hopefully allow capturing from all displays - it does at least on my setup.

However, there are 3 known "glitches":
- sometimes (don't know why), the capture "untriggers" the hi-dpi scaling just during the time the selection is made, causing the capture to be scaled up or down compared to the actual screen. However, even in that situation, area capture should now follow the mouse cursor and select the desired area
- the captured image may look blurry (it's not as bad as in the capture preview though, please save the capture and check it in an image viewer to have an idea of the actual quality),
- sometimes a "clear" area appears outside of the selection being made, but it's harmless

Thanks (or due) to this issue, I really discovered how complex this "hi-dpi" topic is. Windows' handling has changed over time, with now 4 different levels of "hi-dpi awareness (none, global, per monitor and per monitor v2" supported, applications have to declare the way they intend to handle it, then add the Java layer on top of that (with a handling that has evolved with java versions), and then its own set of bugs (see [above](https://github.com/Ginj-capture/Ginj/issues/15#issuecomment-691723376)).

It's interesting to note that the original Jing also struggles in certain situations, and funny to observe how dragging windows (e.g. the Windows Calculator) from a scaled window to a non-scaled one (or reverse) "switches", with the window size suddenly "jumping" and blurry areas appearing on one side.

Ginj v0.4.2 is in no way perfect, but I think it will be "good enough" for the time being.

Don't hesitate to share your experience though. I'm leaving the issue open for now but I intend to close it and open a separate one if only the above glitches are confirmed.

**Comment by janborup** [comment]
**Hi @vicnevicne** 

As mentiontioned [above](https://github.com/Ginj-capture/Ginj/issues/15#issuecomment-678795943)... the Video Capture on a scaled monitor.. is still capturing offset.
https://www.dropbox.com/s/7i0lvlshyhg6mg1/VideoCapture..%200.4.3..125%25%20scaled%20monitor.mp4?dl=0

This is low priority task... 👌

**Comment by vicnevicne** [comment]
Damn :-( 
Found a PC where it still isn't OK :-(
What a dodgy issue...

**Comment by janborup** [comment]
Yes.... I also have same problems....

Normal screen:
![image](https://user-images.githubusercontent.com/6866443/93884435-32e70c80-fce3-11ea-8124-d55f3189e5e4.png)

Ginj Start capture 125%-100%-100% monitors:
![image](https://user-images.githubusercontent.com/6866443/93884599-66299b80-fce3-11ea-9c3d-e79ea18dd6d5.png)

Ginj Start capture 100%-100%-125% monitors:
This can't even take a capture of the left screen!!
![image](https://user-images.githubusercontent.com/6866443/93884806-aab53700-fce3-11ea-9fd0-bb023ca00467.png)
